README.md
# holka-garage

IoT project running on a Raspberry Pi to open/close a garage door and to see if it's opened or closed.

Features:

* Ability to open close garage via website
* Send notifications via IFTTT/Texting when garage opens/closes
* MMS photo sent when garage open/closed clicked

Coming soon:

* Open garage via GPS
* Wire diagram

Required for setup:

1.Raspberry Pi with 16GB+ class 10 SD card  (<https://www.amazon.com/CanaKit-Raspberry-Complete-Starter-Kit/dp/B01C6Q2GSY/ref=sr_1_3?ie=UTF8&qid=1508849215&sr=8-3&keywords=pi+3+kit)>

2.Power relay (<https://www.amazon.com/Indicator-Light-Channel-Module-Arduino/dp/B00P7QDJD2/ref=sr_1_4?s=electronics&ie=UTF8&qid=1508849168&sr=1-4&keywords=power+relay+single+simple)>

3.Optional PIR motion sensor (<https://www.amazon.com/Gowoops-HC-SR501-Pyroelectric-Infrared-Detector/dp/B01MZIQZG1/ref=sr_1_5?ie=UTF8&qid=1508849206&sr=8-5&keywords=pir+motion+sensor)>

4.Pi Camera (<https://www.amazon.com/Arducam-Megapixels-Sensor-OV5647-Raspberry/dp/B012V1HEP4/ref=sr_1_6?ie=UTF8&qid=1508849318&sr=8-6&keywords=raspberry+pi+camera+module>)

5.Door magnent sensor (<https://www.amazon.com/Gikfun-Sensor-Magnetic-Switch-Arduino/dp/B0154PTDFI/ref=sr_1_1_sspa?ie=UTF8&qid=1508849795&sr=8-1-spons&keywords=wired+door+sensor&psc=1>)

6.Optional Hue light bulbs (will need bridge too) (<https://www.amazon.com/Philips-Hue-Equivalent-Dimmable-Assistant/dp/B073SSNNNH/ref=sr_1_4?ie=UTF8&qid=1538656218&sr=8-4&keywords=hue+light+bulbs+white>)

Setup:

1.Setup Rasbperry Pi

2.Get packages installed on RPi <https://github.com/Connor-Knabe/install-scripts>

3.Rename example files in settings folder.  Note the files (login.js, options.js, messengerInfo.js) are .gitignored.  Be sure to NOT commit these files as they contain sensitive info! If you mistype this command you might accidently commit these if you use version control.  Use command below:

    mv optionsExample.js options.js && mv messengerInfoExample.js messengerInfo.js && mv loginExample.js login.js

4.Get a Twilio account (for texting) <https://www.twilio.com/try-twilio>

    4a. Fill out TWILIO_AUTH_TOKEN and TWILIO_ACCOUNT_SID in messengerInfo.js

5.Get an IFTTT account (for push notifications) <https://ifttt.com/join>

    5a.Setup a new applet (webhook, receive a request) with the event name that matches the one in messengerInfoExample.js then notification(garage_open_trigger, garage_alert, etc.)
    5b. Fill out the text such as: {{Value1}} Garage Door Opened or {{Value1}} Garage Door has been open for more than: {{Value2}} minutes!.
    Value1 = iftttValue1 from the messengerInfoExample.js file
    Value2 = mins garage has been opened before alert sends.
    5c.Download and login to the IFTTT mobile app on your phone and turn notifications on.

6.Add your IFTTT API key from <https://ifttt.com/services/maker_webhooks/settings> to the messengerInfo file under iftttRecipients -> ApiKey

7.Clone holka-garage repo <https://github.com/Connor-Knabe/holka-garage>

8.Fill out information including API keys phone numbers etc in the various setting files

9.Use pm2 to start the program using the following command
    sudo pm2 start app.js --name hg && pm2 startup && pm2 --save

10.Setup a domain name and point it to your external ip address.

11.Install ssl certificate <https://github.com/certbot/certbot> use ./certbot-auto certonly --manual --email admin@example.com -d example.com -d www.example.com -d other.example.net note you will need to make sure app.js is running and add a route to routes.js with the specified info to verify this.

11b.If you're using certbot you will want to check to see if you need to renew this script monthly .  You can add a .sh file to /etc/cron.monthly that contains certbot renew --standalone --pre-hook "pm2 stop hg" --post-hook "pm2 start hg" which will run monthly and renew your cert.  Also run sudo chmod +x yourFileName.sh to give permissions to execute it.

12.Change the sslPath variable in login.js to the location of your certificate.

13.Visting https://yourdomain.com should pull up the website.

14.If you've entered geoIpFilter in options.js it will limit ip addresses within those state abbreviations (comma separated) to open the garage door.

15.If you've entered values for vpnIp or localIp it will only allow an ip address from those ip's to open the garage door


changelog.txt
v3.4 - will show when next time garage is estimated to open
v3.3 - using ifttt to turn on/off garage lights
v3.2b - now logs properly when garage attempted to open
v3.1 - tracking time of day garage opens
v2.8.0 - adding ability to disable odd time open alert
v2.7.3-1.0.1 - fixing issue if sets to home before open garage attempt
v2.6.8 - shows what time the light automation will resume
v2.6.7 - shows the toggle for automated hue lights on site
v2.6.6 - section on website that mentions if garage should open based on rules or not
v2.6.5 - adding route to disable lights temporarily
v2.6.4 - adding route to temporarily disable motion sensor and schedule
v2.6.3 - estimate on lift of spring left
v2.6.2 - deployed logging # of garage opens
v2.6.1 - logging # of garage opens to keep track of health of garage spring
v2.5.5 - turning outlets off if away from home if they are on a timer and turned on since homeowners went away
v2.5.4 - alerting users when guest mode is enabled/disabled
v2.5.2 - if opened/closed via website don't overwrite, sending ifttt alert when still open alert is active 
v2.5.1 - add to website who last opened the garage
v2.5.0 - adding guest is home feature
v2.4.3 - refactoring homeAway to move all logic there instead of some in iot.js
v2.4.2 - adding button to disable garage door still open alert
v2.4.1 - adding garage door opened for time
v2.4.0 - major refactor of auth routes
v2.3.0 - fixing security vulnerabilites with unsecured routes
v2.2.0 - adding cron to check for away every hour
v2.1.1 - using home/away to alert if garage opened when no one home
v2.1.0 - changing up alerts for when homeowners are not home
v2.0.8 - not sending response back
v2.0.7 - fixing bug with website timing out and failing health check
v2.0.5 - fixed away time if over 24 hours
v2.0.3 - buttons to manually set home/away for each user
v2.0.0 - calculates time since reboot and ability to set IFTTT to home
v1.9.1 - now shows time since home/away 
v1.9.0 - website to show home/away/timer status for all users
v1.8.2 - text alert for garage being open 30 mins
v1.8.1 - cleaning up debug logs
v1.8.0 - fixing bug with away timer
v1.7.1 - refactoring route file
v1.7.0 - working on setting timer with gps away request
v1.6.6 - refactoring the garage open rules and security alert
v1.6.5 - using correct verbiage in log
v1.6.1 - fix bug that was preventing garage from closing
v1.6.0 - only open garage if you've been gone for 15 minutes to prevent false positives
v1.5.4 - fixing bug with garage open alert
v1.5.3 - disabling tls 1.0 and 1.1
v1.5.1 - adding feature to call if garage still open after first alert
v1.5.0 - updating packages
v1.4.4 - disable gps open button
v1.4.3 - not sending mms when open garage via button
v1.4.0 - adding home/away logic for ifttt to replace nest's home/away 
v1.3.0 - not opening garage until atleast 11am automatically
v1.2.9 - updating npm packages
v1.2.8 - adding check to open garage via GPS on weekend
v1.2.4 - sending generic alert around gps location
v1.2.3 - support for 2nd person to open garage via gps, adjusting time garage will autoopen
v1.2.2 - can open garage via ifttt webhook using gps location from the app
v1.2.1 - fixing bug with reqParams for enabling garage alert
v1.2.0 - adding ability to have two people on gps and alert if neither are home
v1.1.9 - adding favicon to prevent error on loading proxied page
v1.1.8 - removing cert cron, doing it at the server level because it needs to stop the server while running
v1.1.7 - bug fix in total monthly usage calculation
v1.1.6 - adding gps tracking to disable alert when garage open for certain people, brightness different for different situations
v1.1.5 - motion sensor turns garage lights on if hue is enabled 
v1.1.4 - cleaning up logs and adding more feature toggles
v1.1.3 - deleting images from Twilio after sent, adding hueEnergyUsage to reverse proxy
v1.1.2 - refactoring video service
v1.1.1 - added light brightness setting
v1.1.0 - working on using gifs instead of jpegs, also deleting mms after sent with twilio
v1.0.7 - taking photo before sending alert after garage has been opened for a while
v1.0.6 - updating readme
v1.0.5 - adding geoip lookup
v1.0.4 - adding alert if garage open for more than a certain time
v1.0.3 - major refactor, added options file, formatted files using prettier, adding debug to option.js file
v1.0.2 - hue integration option to use IFTTT for alerts
v0.4 - refactoring
v0.3 - added secure cookies and ssl


services/iot.js
const options = require('../settings/options.js');

var motionSensor = null;
var garageSensor = null;
var garageSwitch = null;


if(!options.localDebug){
	const Gpio = require('onoff').Gpio;
	motionSensor = new Gpio(15, 'in', 'both');
	garageSensor = new Gpio(4, 'in', 'both');
	garageSwitch = new Gpio(21, 'out');
} 


var motionSensorTimeoutOne = null,
	motionSensorTimeoutTwo = null,
	hasTurnedLightsOn = false,
	garageOpenAlertOneTimeout = null,
	garageOpenAlertTwoTimeout = null,
	expectedGarageOpen = false,
	manualGarageOpenTimeout = null,
	personOneShouldOpenTimer = false,
	personTwoShouldOpenTimer = false,
	personOneShouldOpenTimerTimeout = null,
	personTwoShouldOpenTimerTimeout = null,
	garageLastOpenedTime = null,
	garageLastClosedTime = null,
	temporaryDisableGarageStillOpenAlert = false,
	tempGarageDisableStillOpenAlertTimeout = null,
	temporaryDisableGarageStillOpenAlertTime = new Date(),
	temporaryEnableGuestIsHomeTimeout = null,
	temporaryEnableGuestIsHomeTime = null,
	shouldAlertTimeoutOne = null,
	shouldAlertTimeoutTwo = null;

module.exports = function(app, debugMode, io, logger, video, messenger, hue, cron, homeAway,garageTimeRules,garageTracking) {
	var Status = {
		wasOpenedViaWebsite: false,
		wasClosedViaWebsite: false
	};

	var hasBeenOpened = garageIsOpen();
	const messengerInfo = require('../settings/messengerInfo.js');

	const fs = require("fs").promises;

	app.set('takingVideo', false);
	if(!options.localDebug){
		garageSensor.watch(async function(err, value) {
			if (err) {
				logger.error('Error watching garage sensor: ', err);
			}
			if (value == 1 && !hasBeenOpened) {
				io.sockets.emit('garageStatus', 'open');
				garageTracking.garageOpens++;
				io.sockets.emit('garageOpenCount', getGarageOpenCount());
				io.sockets.emit('springLifeRemaining', getSpringLifeRemaining());
				shouldAlertHomeOwnersBasedOnTime('opened');
				
				await writeToGarageTrackingFile();
	
				hasBeenOpened = true;
				garageLastOpenedTime = new Date();
				io.sockets.emit('garageLastOpenedTime', garageLastOpenedTime);
	
				var msg = 'Garage door opened';
	
				if (options.enableLightsOnGarageOpen) {
					hue.garageLightsOnTimed();
				}
				logger.debug(`garage open {Status.wasOpenedViaWebsite} ${Status.wasOpenedViaWebsite}`);
	
				garageAlertStillOpenCheck(options.garageOpenAlertOneMins, garageOpenAlertOneTimeout, false);
				clearTimeout(shouldAlertTimeoutOne)
				clearTimeout(shouldAlertTimeoutTwo)
				shouldAlertTimeoutOne = setTimeout(() => {
					shouldAlertHomeOwners('opened');
				}, 30 * 1000);
	
				logger.debug(msg);
				io.sockets.emit('garageErrorStatus', null);
				const open = true;
				if(!Status.wasOpenedViaWebsite){
					io.sockets.emit('whoOpenedGarageLast', homeAway.Status.getWhoJustOpenedOrClosedGarage(open,true));
				}
				
			} else if (value == 0 && hasBeenOpened) {
				shouldAlertHomeOwnersBasedOnTime('closed');
	
				clearTimeout(tempGarageDisableStillOpenAlertTimeout);
				temporaryDisableGarageStillOpenAlert = false;
				io.sockets.emit('garageStatus', 'closed');
	
				hasBeenOpened = false;
				garageLastClosedTime = new Date();
				io.sockets.emit('garageLastClosedTime', garageLastClosedTime);
	
				var msg = 'Garage door closed';
				clearTimeout(garageOpenAlertOneTimeout);
				clearTimeout(garageOpenAlertTwoTimeout);
				clearTimeout(shouldAlertTimeoutOne)
				clearTimeout(shouldAlertTimeoutTwo)
				shouldAlertTimeoutTwo = setTimeout(() => {
					shouldAlertHomeOwners('closed');
				}, 30 * 1000);4
	
				logger.debug(`Garage door closed {Status.wasClosedViaWebsite} ${Status.wasClosedViaWebsite}`);
	
				io.sockets.emit('garageErrorStatus', null);
				const open = false;
				if(!Status.wasClosedViaWebsite){
					io.sockets.emit('whoClosedGarageLast', homeAway.Status.getWhoJustOpenedOrClosedGarage(open,true));
				}
			}
		});
	}
	

	//if motion is detected in garage turn on hue lights
	if (!options.localDebug && options.enableMotionSensor && options.enableHue) {
		motionSensor.watch((err, value) => {
			if (err) {
				logger.error('Error watching motion sensor: ', err);
			}
			if (value == 1 && !hasTurnedLightsOn) {
				clearTimeout(motionSensorTimeoutOne);
				motionSensorTimeoutOne = setTimeout(function() {
					hasTurnedLightsOn = true;
				}, 5 * 1000);
				hue.garageLightsOnTimed();
			} else if (value == 0 && hasTurnedLightsOn) {
				clearTimeout(motionSensorTimeoutTwo);
				motionSensorTimeoutTwo = setTimeout(function() {
					hasTurnedLightsOn = false;
				}, 5 * 1000);
			}
		});
	}

	function shouldAlertHomeOwnersBasedOnTime(status){
		logger.debug('Should check alert homeowners based on odd time');

		if(garageTimeRules.shouldAlertBasedOnTime()){
			logger.debug('Should alert homeowners based on odd time');
			if(status=="opened"){
				messenger.sendCallAlert();
			}
			video
			?.streamVideo()
			.then(() => {
				var garageAlertMsg = `ALERT! The garage has been ${status} during odd hours!`;
				messenger.send(true, messengerInfo.toNumbers, garageAlertMsg, sendPictureText, true);
				video?.stopStreaming();
			})
			.catch(() => {
				var garageAlertMsg = `ALERT! The garage has been ${status} during odd hours`;
				sendPictureText = false;
				messenger.send(true, messengerInfo.toNumbers, garageAlertMsg, sendPictureText, true);
				video?.stopStreaming();
			});
		}


	}

	function shouldAlertHomeOwners(status) {
		logger.debug(`possibly alert home owners not home? expectedGarageOpen${expectedGarageOpen} | homeAway.Status.personOneAway && homeAway.Status.personTwoAway ${homeAway.Status.personOneAway} && ${homeAway.Status.personTwoAway}`);
		if (!expectedGarageOpen) {
			if (homeAway.Status.personOneAway && homeAway.Status.personTwoAway && options.shouldAlertIfBothOwnersAwayAndOpen) {
				var sendPictureText = true;
				messenger.sendCallAlert();
				video
					?.streamVideo()
					.then(() => {
						var garageAlertMsg = `The garage has been ${status} but the homeowners are not home!`;
						messenger.send(true, messengerInfo.toNumbers, garageAlertMsg, sendPictureText, true);
						video?.stopStreaming();
					})
					.catch(() => {
						var garageAlertMsg = `The garage has been ${status} but the homeowners are not home! Error taking new video.`;
						sendPictureText = false;
						messenger.send(true, messengerInfo.toNumbers, garageAlertMsg, sendPictureText, true);
						video?.stopStreaming();
					});
			}
		}
	}

	function getGarageOpenCount(){
		return garageTracking.garageOpens;
	}

	function getSpringLifeRemaining(){

		var startDate = new Date(garageTracking.springReplacedDate);
		var currentDate = new Date();
		
		var daysSinceNewSpringInstalled = (currentDate.getTime() - startDate.getTime()) / (1000 * 3600 * 24); 

		const maxCycles = 10000;
		
		var timesOpenedPerDay = garageTracking.garageOpens/daysSinceNewSpringInstalled;

		var cyclesLeft = maxCycles-garageTracking.garageOpens;
		var daysTillNewSpring = cyclesLeft/timesOpenedPerDay;
		var yearsTillNewSpring = daysTillNewSpring/365;

		var years = yearsTillNewSpring;
		var months = (yearsTillNewSpring % 1)*12;
		var days = (months % 1)*31;

		return `${Math.floor(years)} years, ${Math.floor(months)} months, ${Math.floor(days)} days`;
	}

	function garageAlertStillOpenCheck(timeUntilAlert, timeOut, shouldCall) {
		clearTimeout(timeOut);
		timeOut = setTimeout(() => {
			if (garageIsOpen() && !temporaryDisableGarageStillOpenAlert) {
				setTimeout(() => {
					garageAlert(timeUntilAlert, shouldCall);
				}, 45 * 1000);
			}
		}, timeUntilAlert * 60 * 1000);
	}

	function garageIsOpen() {
		var isOpen = false;

		if(!options.localDebug){
			isOpen = garageSensor.readSync() == 1 ? true : false;
		}
		return isOpen;
	}

	function garageAlert(timeUntilAlert, shouldCall) {
		var garageAlertOpenMins = shouldCall ? timeUntilAlert + options.garageOpenAlertOneMins : timeUntilAlert
		var garageAlertMsg = `Garage has been open for more than: ${garageAlertOpenMins} minutes!`;
		logger.debug(garageAlertMsg);
		if (options.garageOpenMinsAlert) {
			logger.debug(garageAlertMsg);
			video
				?.streamVideo()
				.then(() => {
					if (shouldCall) {
						messenger.sendCallAlert();
					}
					messenger.send(options.alertButtonPressTexts, messengerInfo.toNumbers, garageAlertMsg, options.alertSendPictureText, true);
					video?.stopStreaming();
				})
				.catch(() => {
					if (shouldCall) {
						messenger.sendCallAlert();
					}
					messenger.send(options.alertButtonPressTexts, messengerInfo.toNumbers, garageAlertMsg, options.alertSendPictureText, true);
					video?.stopStreaming();
				});
			if (!shouldCall) {
				garageAlertStillOpenCheck(options.garageOpenAlertTwoMins, garageOpenAlertTwoTimeout, true);
			}
		}
		messenger.sendIfttGarageOpenedAlert(options.iftttSendGarageOpenAlert, timeUntilAlert);
	}

	async function toggleGarageDoor(gpsPerson, remoteAddress) {
		if (shouldOpenGarageBaesdOnRules()) {
			if (!garageIsOpen()) {
				logger.info(`Opening garage via gps person ${gpsPerson} from ip: ${remoteAddress}`);
				openCloseGarageDoor();
				messenger.sendIftt(true, `Garage open via GPS for person ${gpsPerson}`);
			} else {
				logger.info(`Attempted to open garage via gps person ${gpsPerson} from ip: ${remoteAddress} but garage was open`);
			}
		} else {
			messenger.sendIftt(true, `Not opening for person ${gpsPerson} due to time range`);
			logger.info(`Not opening garage for person ${gpsPerson} outside of time range from ip: ${remoteAddress}`);
		}
		await writeToGarageTrackingFile();

	}

	function shouldOpenGarageBaesdOnRules(){
		// var shouldOpenGarage = false;
		const shouldOpenGarageBasedOnDayTimeLog = garageTimeRules.shouldOpenCheckAndLog(garageTracking.garageTimesToOpenLog);
		const shouldOpenGarageBasedOnDayTime = garageTimeRules.shouldOpenCheck(garageTracking.garageTimesToOpen);
		
		shouldOpenGarage = shouldOpenGarageBasedOnDayTime || shouldOpenGarageBasedOnDayTimeLog;
		return shouldOpenGarage;
	}

	function openCloseGarageDoor() {
		expectedGarageOpen = true;
		clearTimeout(manualGarageOpenTimeout);
		manualGarageOpenTimeout = setTimeout(() => {
			expectedGarageOpen = false;
		}, 60 * 1000);

		if (!debugMode && !options.localDebug) {
			logger.debug('Opening/closing door now');
			garageSwitch.writeSync(1);
			setTimeout(function() {
				garageSwitch.writeSync(0);
			}, 1000);
		}
	}

	function activateGarageGpsOpenAwayTimer(personTwo) {
		if (personTwo) {
			clearTimeout(personTwoShouldOpenTimerTimeout);
			logger.debug(`personTwoTimer QUEUED for ${options.minsToWaitAfterLeavingHouseForGPSOpen}`);
			personTwoShouldOpenTimerTimeout = setTimeout(() => {
				personTwoShouldOpenTimer = true;
				logger.debug('personTwoTimer Active');
			}, options.minsToWaitAfterLeavingHouseForGPSOpen * 60 * 1000);
		} else {
			clearTimeout(personOneShouldOpenTimerTimeout);
			logger.debug(`personOneTimer QUEUED for ${options.minsToWaitAfterLeavingHouseForGPSOpen}`);
			personOneShouldOpenTimerTimeout = setTimeout(() => {
				personOneShouldOpenTimer = true;
				logger.debug('personOneTimer Active');
			}, options.minsToWaitAfterLeavingHouseForGPSOpen * 60 * 1000);
		}
	}

	async function garageDoorOpenHandler(isPersonTwo, gpsPerson, remoteAddress) {
		var personTimerShouldOpen = isPersonTwo ? personTwoShouldOpenTimer : personOneShouldOpenTimer;
		if (!personTimerShouldOpen) {
			const logMsg = `Not opening for person ${gpsPerson} due to timer. ${options.minsToWaitAfterLeavingHouseForGPSOpen} min delay`;
			logger.debug(logMsg);
			messenger.sendGenericIfttt(logMsg);
		} else {
			logger.debug('should open garage');
			if (isPersonTwo) {
				clearTimeout(personTwoShouldOpenTimerTimeout);
				personTwoShouldOpenTimer = false;
			} else {
				clearTimeout(personOneShouldOpenTimerTimeout);
				personOneShouldOpenTimer = false;
			}
			logger.debug('toggle garage door');

			// await garageTimeRules.shouldLog(garageTracking.garageTimesToOpenLog);
			// await writeToGarageTrackingFile();
			toggleGarageDoor(gpsPerson, remoteAddress);
		}
	}

	function getGarageLastOpenedTime() {
		return garageLastOpenedTime;
	}

	function getGarageLastClosedTime() {
		return garageLastClosedTime;
	}

	//clean up these two functions 
	function getTemporaryGuestIsHomeStatus(){

		var guestHomeTime = "";
		if(temporaryEnableGuestIsHomeTime){
			guestHomeTime = `until ${temporaryEnableGuestIsHomeTime.toLocaleTimeString()}`;
		}
		var status = homeAway.Status.temporaryEnableGuestIsHome ? `Guest Is Home Enabled ${guestHomeTime}` : `Guest Is Home Disabled`;

		return status
	}

	function toggleTemporaryEnableGuestIsHome(){
		if(homeAway.Status.temporaryEnableGuestIsHome){
			homeAway.Status.temporaryEnableGuestIsHome = false;
			messenger.sendGenericIfttt(`Guest is NOT home shutting off lights when home owners are away`);
			clearTimeout(temporaryEnableGuestIsHomeTimeout);
		} else {
			homeAway.Status.temporaryEnableGuestIsHome = true;
			temporaryEnableGuestIsHomeTime = new Date(new Date().setHours(new Date().getHours() + options.guestIsHomeEnableForHours));
			messenger.sendGenericIfttt(`Guest is home until ${temporaryEnableGuestIsHomeTime} NOT shutting off lights when home owners are away`);
			temporaryEnableGuestIsHomeTimeout = setTimeout(()=>{
				homeAway.Status.temporaryEnableGuestIsHome = false;
			},options.guestIsHomeEnableForHours*60*60*1000)
		}

		return getTemporaryGuestIsHomeStatus();
	}

	function getTemporaryDisableGarageStillOpenAlertStatus(){
		var status = temporaryDisableGarageStillOpenAlert ? `Still Open Alert Disabled until ${temporaryDisableGarageStillOpenAlertTime.toLocaleTimeString()}` : 'Still Open Alert Enabled';

		return status
	}

	function toggleTemporaryDisableGarageStillOpenAlert(){
		if(temporaryDisableGarageStillOpenAlert){
			messenger.sendGenericIfttt(`Alerting users when garage has been left open`);
			temporaryDisableGarageStillOpenAlert = false;
			clearTimeout(tempGarageDisableStillOpenAlertTimeout);
		} else {
			temporaryDisableGarageStillOpenAlert = true;
			temporaryDisableGarageStillOpenAlertTime = new Date(new Date().setHours(new Date().getHours() + options.garageStillOpenAlertDisableForHours));
			messenger.sendGenericIfttt(`NOT alerting users when garage has been left open until ${temporaryDisableGarageStillOpenAlertTime}`);

			tempGarageDisableStillOpenAlertTimeout = setTimeout(()=>{
				temporaryDisableGarageStillOpenAlert = false;
			},options.garageStillOpenAlertDisableForHours*60*60*1000)
		}

		return getTemporaryDisableGarageStillOpenAlertStatus();
	}

	//cron
	var job = new cron(
		'5 * * * *',
		function() {
			if (homeAway.Status.isAway()) {
				//turn off specific iot devices that may be on a schedule to turn on at top of the hour
				messenger.sendIftt(null, 'set away', messengerInfo.iftttGarageSetAwayUrl);
			}
		},
		null,
		true,
		'America/Chicago'
	);
	job.start();

	async function writeToGarageTrackingFile() {
		try { 
			await fs.writeFile( "garageTracking.json", JSON.stringify( garageTracking ), "utf8");
		} catch(err) { 
			logger.error(`Error writing garageTracking file ${err}`); 
		} 	
	}

	function getGarageStatus(){
		var garageGPSOpenTime = garageTimeRules.nextOpenBasedOnDayTime();
		const time = new Date(garageGPSOpenTime).toLocaleTimeString('en-US',{ hour12: true, hour: 'numeric'});
		const date = new Date(garageGPSOpenTime).toLocaleDateString(undefined,{month: 'numeric', day: 'numeric'});
		garageGPSOpenTime = `${time} ${date}`
		logger.debug("garageGPSOpenTime",garageGPSOpenTime);

		const shouldOpen = shouldOpenGarageBaesdOnRules() ? "Y" : garageGPSOpenTime;

		var garageOpenClosed = garageIsOpen() ? "O": "C";
		var personOneAway = homeAway.isPersonAway(false) ? "A": "H";
		var personTwoAway = homeAway.isPersonAway(true) ? "A": "H";
		
		return `${garageOpenClosed}|1${personOneAway}|2${personTwoAway}|${shouldOpen}`;
	}

	return {
		garageIsOpen: garageIsOpen,
		toggleGarageDoor: toggleGarageDoor,
		activateGarageGpsOpenAwayTimer: activateGarageGpsOpenAwayTimer,
		garageDoorOpenHandler: garageDoorOpenHandler,
		openCloseGarageDoor: openCloseGarageDoor,
		getGarageLastOpenedTime: getGarageLastOpenedTime,
		getGarageLastClosedTime:getGarageLastClosedTime,
		toggleTemporaryDisableGarageStillOpenAlert:toggleTemporaryDisableGarageStillOpenAlert,
		getTemporaryDisableGarageStillOpenAlertStatus:getTemporaryDisableGarageStillOpenAlertStatus,
		toggleTemporaryEnableGuestIsHome:toggleTemporaryEnableGuestIsHome,
		getTemporaryGuestIsHomeStatus:getTemporaryGuestIsHomeStatus,
		Status:Status,
		getGarageOpenCount:getGarageOpenCount,
		getSpringLifeRemaining:getSpringLifeRemaining,
		shouldOpenGarageBaesdOnRules:shouldOpenGarageBaesdOnRules,
		getGarageStatus:getGarageStatus
	};
};
